sympy/core/tests/test_expr.py

```diff
@@ -1414,6 +1414,9 @@ def test_coeff():
     assert (n*m + o*m*n).coeff(m*n) == o
     assert (n*m + o*m*n).coeff(m*n, right=1) == 1
     assert (n*m + n*m*n).coeff(n*m, right=1) == 1 + n  # = n*m*(n + 1)
+    assert (x*n + y*n + z*m).coeff(n) == x + y #i issue #16752
+    assert (n*m + n*o + o*l).coeff(n, right=1) == m + o
+    assert (x*n*m*n + y*n*m*o + z*l).coeff(m, right=1) == x*n + y*o
```

Comment:
I think we need an assurance the coefficient of the left and right will recreate the terms that were matched:

```
>>> (x*n*m*n + y*n*m*o + z*l).coeff(m, right=0)
n
>>> (x*n*m*n + y*n*m*o + z*l).coeff(m, right=1)
x*n + y*o
```
That's right because `factor_nc(x*n*m*n + y*n*m*o) == n*m*(x*n + y*o)`. But consider the following:
```python
>>> (x*n*m*n + x*n*m*o + z*l).coeff(m, right=0)
x*n
>>> (x*n*m*n + x*n*m*o + z*l).coeff(m, right=1)
x*n + x*o
```
The `x` cannot be part of the left and part of the right.